### DISCUSSION TITLE
Web Game/Mod Editor

### DISCUSSION THREAD

**Discussion by zicklag** [fishfight/punchy#256] [Ideas]
With our Punchy webplayer, people can already create their own GitHub repos and use GitHub pages to host their own modified game assets. This means we could create a template GitHub repo that anybody could fork, modify, and instantly get a link that they could use to share with friends and show them their custom Punchy mod, all without having to install anything on their computer.

This is a really cool workflow, but I think we can take it to even another level.

A growing trend is to allow people to contribute to GitHub projects with web workspaces such as the new GitHub Codespaces, which gives you a VSCode environment in your browser with dev tools built-in. To supercharge the Punchy mod development workflow, we can make a web-compatible VSCode plugin that embeds the Punchy web player that features hot reload and running the game right inside a VSCode tab, while being able to edit the Punchy mods in the same VSCode web instance. You could have a full-blown Punchy mods IDE right in your browser with instant feedback and hot reload!

We could even integrate a visual programming language such as Blockly into the plugin! ( #257 )

T user experience could be like this ( I think ):

1. Go to the Punchy Mod GitHub template repo
    - There will be a nice, informative README with instructions on how to use the repo.
2. Click the create repo from template button in GitHub
    - Now you have your own Punchy game!
    - The template repo comes with the GitHub actions installed so that there it will generate a GitHub pages site hosting the game and updating the README with a link to run the forked game in the Punchy webplayer.
    - Now we want to mod our forked game.
3. Click the Codespaces tab next to the clone menu and create a new codespace
    - GitHub will provision and configure a remote VSCode instance that can be accessed from the browser.
    - We can configure our template repo to pre-install our Punchy development plugin
4. Now you can edit any of the the files in your Punchy fork
5. Now you can click the Punchy icon in the left sidebar to open a VSCode tab running the punchy webplayer and connected to the files that you are editing.
6. While the game is running, you could edit files and see them hot reloaded in the web player tab.

---

This would be some work to setup, but at the same time, I don't think it's that far out of reach.

**Comment by erlend-sh** [comment]
Yesss, let’s definitely work towards this. Sounds like a great use of the GitHub platform.
